text/0000-semantic-test-selectors.md

```diff
@@ -0,0 +1,700 @@
+This is perhaps the most visceral motivation. Which test would you rather read?
+
+This one:
+
+```js
+await fillIn('.login-form .field-email', 'alice@example.com');
+await fillIn('.login-form .field-password', 'topsecret');
+await click('.submit-btn');
+```
+
+Or this one:
+
+```js
+await fillIn('Email', 'alice@example.com');
+await fillIn('Password', 'topsecret');
+await click('Log in');
+```
+
+## Detailed design
+
+We will provide semantic variants of `click` and `fillIn`. Their proposed
+behaviours are given below.
+
+### Semantic `click`
+
+```ts
+function click(label: string): Promise<void>
+```
+
+This helper finds a **link** or **button** whose **perceivable text** contains
+the string specified by `label`, or a **switch** whose **perceivable label**
+contains the string specified by `label`.
+
+If it finds exactly one matching element, it invokes the general-purpose
+`click` helper upon it.
+
+If it finds multiple matching elements, it throws:
+
+```
+Found multiple clickable elements with perceivable text '${label}' when calling
+`click('${label}')`. You may want to use a more descriptive label in the test
+and the template.
+```
+
+If it finds no matching element, it throws:
+
+```
+Clickable element with perceivable text '${label}' not found when calling `click('${label}')`.
+```
+
+### Semantic `fillIn`
+
+```ts
+function fillIn(label: string, value: string): Promise<void>
+```
+
+This helper finds an **input** whose **perceivable label** contains the string
+specified by `label`.
+
+If it finds a matching element, it invokes the general-purpose `fillIn` helper
+upon it with `value` as the second argument.
+
+If a label is found without a counterpart form control it throws:
+
+```
+Form control not found for label containing '${label}' when calling
+`fillIn('${label}', '${value}')`. You may want to check the `for` attribute of
+the label.
+```
+
+If no label is found, it throws:
+
+```
+Form control labelled '${label}' not found when calling `fillIn('${label}', '${value}')`.
+If the form control exists, you may want to label its using a label, title or aria-label.
+```
+
+### Definitions
+
+A **link** is one of the following:
+
+- `a`
+- `[role="link"]`
+- `[role="menuitem"]`
+
+A **button** is one of the following:
+
+- `button`
+- `input[type="button"]`
+- `input[type="reset"]`
+- `input[type="submit"]`
+- `[role="button"]`
+
+An **input** is one of the following:
+
+- `input`
+- `textarea`
+- `select`
+- `[role="slider"]`
+- `[role="spinbutton"]`
+- `[role="textbox"]`
+- `[contenteditable="true"]`
+
+A **switch** is one of the following:
+
+- `input[type="checkbox"]`
+- `input[type="radio"]`
+- `[role="checkbox"]`
+- `[role="option"]`
+- `[role="radio"]`
+
+A **form control** is an **input** or **switch**.
+
+The **perceivable text** of a **link**, **button**, or other element may be
+found in any/all of the following locations **if-and-only-if the element is
+perceivable**:
+
+- Its `innerText` property
+- Its `title` attribute
```

Comment:
it will only perceive title of no other attribute is present

Reply:
@billybonks yes, exactly. Do you feel like lines 188/189 don't clarify that enough?

Reply:
not really because it seems like it is saying that it could use anyone of those attributes and doesn't explain what a perceivable element is. should probably mention a perceivable element is defined here 
https://www.w3.org/TR/accname-1.1/#mapping_additional_nd_te